daemon/endpoint.go

```diff
@@ -1024,3 +1000,51 @@ func (d *Daemon) RemoveRestoredDNSRules(epID uint16) {
 	}
 	proxy.DefaultDNSProxy.RemoveRestoredRules(epID)
 }
+
+// TODO: Review this when I'm not split-brained
+type endpointManager struct {
+	*endpointmanager.EndpointManager
+
+	d *Daemon
+}
```

Comment:
Maybe would be better to wrap this up with a wrapper of multiple interfaces, like `pkg/endpoint/endpoing.go#ipReleaser`, and `pkg/endpoint/endpoing.go#monitorOwner` ?

Reply:
I don't think it makes a difference, since `daemon/` has to know about all the concrete types anyway. Abstracting away to use these interfaces is useful in `pkg/` as we can deal with abstract types and not deal directly with the concrete types; or for testing where mocking is important to allow better testing of some core functionality. Neither of those reasons applies here.

Anyhow, this is just a minimal patch to expose the functionality we need. Looking a little more, it seems to me that `endpoint.Delete` should really be `endpointmanager.Delete(ep)` and then `endpointmanager` should have the `ipReleaser` / `monitorOwner` interfaces and actually manage the endpoints instead of passing an endpointmanager into the endpoint package to allow an endpoint to delete itself. If we did this, then I think we can get rid of this new type as it's just making up for the lack of proper structure in the code below it. Maybe there's a safe way to refactor this now but it seems like a better target for the master branch code development rather than a bugfix.

Reply:
^^^ In the latest iteration, this got slightly uglier, but I still didn't want to unravel too many things here. Once we have verified the fix, I plan to follow up on refactoring to clean this up as part of master development of the fix.